### PR TITLE
Deny numpy arrays in ufunc outside fuse decorator

### LINKED ISSUE #147: Deny numpy arrays in ufunc outside fuse decorator
Currently `ufunc`s defined in `cupy.fusion` accepts numpy arrays. This is not intended.

**Expected behavior** (`cupy.math.arithmetic.add`)
```
>>> cupy.math.arithmetic.add(numpy.array(1), numpy.array(2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/core/elementwise.pxi", line 736, in cupy.core.core.ufunc.__call__ (cupy/core/core.cpp:48698)
    args = _preprocess_args(args)
  File "cupy/core/elementwise.pxi", line 110, in cupy.core.core._preprocess_args (cupy/core/core.cpp:36647)
    raise TypeError('Unsupported type %s' % type(arg))
TypeError: Unsupported type <class 'numpy.ndarray'>
```

**Fusion** (`cupy.add` = `cupy.fusion.add`)
```
>>> cupy.fusion.add(numpy.array(1), numpy.array(2))
3
```

These `ufunc`s should accept `numpy.ndarray` only within `@fuse` decorator.

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #147 

(Additional notes)
I removed all numpy-cupy comparison tests in `test_fusion.py`.
Instead, equality is tested between non-fused cupy computation and fused cupy computation.

### CODE REVIEW

File: cupy/core/fusion.py [hunk 1]
```diff
@@ -14,6 +15,9 @@
 from cupy import util
 
 
+_thread_local = threading.local()
+
```
Comment (okuta) [inline cupy/core/fusion.py]: How about adding `_thread_local.in_fusion = False`?

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 2]
```diff
@@ -554,24 +554,16 @@ def f(*x):
...
-        cret = list(cret) if type(cret) == tuple else [cret]
-        fcret = list(fcret) if type(fcret) == tuple else [fcret]
-        for n, c, fc in zip(nret, cret, fcret):
-            numpy.testing.assert_array_almost_equal(n, c.get())
-            numpy.testing.assert_array_almost_equal(n, fc.get())
+            data = [gen(*args) for i in range(n)]
+
+        ret0 = func(*data)  # Non-fused
+        ret1 = f(*data)  # Fused
+        ret0 = list(ret0) if isinstance(ret0, tuple) else [ret0]
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_fusion.py]: Please remove `list` casting.

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 3]
```diff
@@ -554,24 +554,16 @@ def f(*x):
...
+        ret1 = list(ret1) if isinstance(ret1, tuple) else [ret1]
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_fusion.py]: ditto

File: cupy/core/fusion.py [hunk 4]
```diff
@@ -684,12 +695,14 @@ def __repr__(self):
         return repr(self._cupy_op)
 
     def __call__(self, *args, **kwargs):
-        if builtins.any(type(_) is _FusionRef for _ in args):
-            return _convert(self._fusion_op)(*args, **kwargs)
-        elif builtins.any(type(_) is numpy.ndarray for _ in args):
-            return self._numpy_op(*args, **kwargs)
-        else:
-            return self._cupy_op(*args, **kwargs)
+        in_fusion = getattr(_thread_local, 'in_fusion', False)
```
Comment (okuta) [inline cupy/core/fusion.py]: How about using `in_fusion = _thread_local.in_fusion`?

File: cupy/core/fusion.py (hunk 1 above)
Comment (niboshi) [inline cupy/core/fusion.py]: In my understanding, you cannot set a default value like this, right?

Comment (okuta) [inline cupy/core/fusion.py]: Does Python copy `_thread_local` when we create new thread?

File: cupy/core/fusion.py (hunk 4 above)
Comment (okuta) [inline cupy/core/fusion.py]: This is my mistake.

**Comment by okuta** [comment]
LGTM!

File: cupy/core/fusion.py (hunk 1 above)
Comment (niboshi) [inline cupy/core/fusion.py]: Looks like no.
```
import threading

_tls = threading.local()
_tls.foo = 4

def f(n):
    try:
        print(n, _tls.foo)
    except AttributeError:
        print(n, 'AttributeError')
    _tls.foo = 10
def func(n):
    f(n)
    f(n)
    f(n)

threads = [threading.Thread(target=func, args=(_,)) for _ in range(3)]
for t in threads: t.start()
for t in threads: t.join()
```

```
0 AttributeError
0 10
1 AttributeError
0 10
1 10
2 AttributeError
1 10
2 10
2 10
```

Comment (okuta) [inline cupy/core/fusion.py]: Thanks, I understand.

**Pull request merged**